compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/PackToIntrinsics.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUDialect.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.h"
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Passes.h"
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::IREE::GPU {
+
+#define GEN_PASS_DEF_PACKTOINTRINSICSPASS
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Passes.h.inc"
+
+namespace {
+struct PackToIntrinsicsPass final
+    : impl::PackToIntrinsicsPassBase<PackToIntrinsicsPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+LogicalResult packToIntrinsic(linalg::LinalgOp linalgOp,
+                              RewriterBase &rewriter) {
+  auto loweringConfig =
+      getLoweringConfig<IREE::GPU::LoweringConfigAttr>(linalgOp);
+  if (!loweringConfig) {
+    return success();
+  }
+
+  IREE::GPU::MmaInterfaceAttr kind = loweringConfig.getMmaKind();
+  if (!kind) {
+    return success();
+  }
+
+  // At this point fail compilation if packing fails because the operation
+  // is expected to be packed to the intrinsic shape.
```

Comment:
nit: I would move the matching to when you are walking the linalg ops and getting candidates, and then you don't need to return success when it fails to match (can also return notifyMatchFailure then).